### PR TITLE
HIVE-27950: STACK UDTF returns wrong results when # of argument is no…

### PR DESCRIPTION
**Pull request by okumin**
…t a multiple of N

https://issues.apache.org/jira/browse/HIVE-27950

### What changes were proposed in this pull request?

Fix a data integrity problem of the stack UDF.

### Why are the changes needed?

This is a bug fix

### Does this PR introduce _any_ user-facing change?

Yes. This PR would prevent users from facing the issue and would change error messages a bit

### Is the change a dependency upgrade?

No

### How was this patch tested?

I added test cases

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java [hunk 1]
```diff
@@ -63,11 +64,16 @@ public StructObjectInspector initialize(ObjectInspector[] args)
     }
     if (!(args[0] instanceof ConstantObjectInspector)) {
       throw new UDFArgumentException(
+          "The first argument to STACK() must be a constant.");
+    }
+    final Object value = ((ConstantObjectInspector) args[0]).getWritableConstantValue();
+    if (!(value instanceof IntWritable)) {
```
Comment (okumin) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java:70]: I added a validation since the current implementation throws a ClassCastException. I can split PRs if we don't want to fix it in this PR.

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java [hunk 2]
```diff
@@ -63,13 +64,21 @@ public StructObjectInspector initialize(ObjectInspector[] args)
     }
     if (!(args[0] instanceof ConstantObjectInspector)) {
       throw new UDFArgumentException(
+          "The first argument to STACK() must be a constant.");
```
[the remark this answers is not in the document]
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java:67]: The first argument, # of rows to be generated from a single row, defines the final schema. That's why it must be evaluated at compile time.

```
0: jdbc:hive2://hive-hiveserver2:10000/defaul> select stack(2, 'a', 'b', 'c', 'd', 'e', 'f');
...
+-------+-------+-------+
| col0  | col1  | col2  |
+-------+-------+-------+
| a     | b     | c     |
| d     | e     | f     |
+-------+-------+-------+
2 rows selected (3.979 seconds)
0: jdbc:hive2://hive-hiveserver2:10000/defaul> select stack(3, 'a', 'b', 'c', 'd', 'e', 'f');
...
+-------+-------+
| col0  | col1  |
+-------+-------+
| a     | b     |
| c     | d     |
| e     | f     |
+-------+-------+
```

**Comment by InvisibleProgrammer** [comment]
LGTM

**Review by InvisibleProgrammer — APPROVED** [review]

**Comment by okumin** [comment]
@aturoczy @InvisibleProgrammer Do you know anyone who can get this PR approved? Thanks.

**Comment by aturoczy** [comment]
@SourabhBadhya @deniskuzZ @abstractdog Could you pls check this?

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java [hunk 3]
```diff
@@ -63,13 +64,21 @@ public StructObjectInspector initialize(ObjectInspector[] args)
...
+    final Object value = ((ConstantObjectInspector) args[0]).getWritableConstantValue();
+    if (value == null) {
+      throw new UDFArgumentException("The first argument of STACK() must not be null.");
+    }
+    if (!(value instanceof IntWritable)) {
+      throw new UDFArgumentTypeException(
+          0,
           "The first argument to STACK() must be a constant integer (got " +
-          args[0].getTypeName() + " instead).");
+              args[0].getTypeName() + " instead).");
```
Comment (SourabhBadhya) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java:77]: nit: Unnecessary change

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java [hunk 4]
```diff
@@ -109,15 +118,15 @@ public StructObjectInspector initialize(ObjectInspector[] args)
...
+      throws HiveException {
     for (int ii = 0; ii < numRows.get(); ++ii) {
       for (int jj = 0; jj < numCols; ++jj) {
         int index = ii * numCols + jj + 1;
         if (index < args.length) {
           forwardObj[jj] = 
             returnOIResolvers.get(jj).convertIfNecessary(args[index], argOIs.get(index));
         } else {
-          forwardObj[ii] = null;
+          forwardObj[jj] = null;
```
Comment (SourabhBadhya) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java:129]: I think you could have fixed a bug here, is there a valid test case which exposes this bug which can be added here?

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java [hunk 5]
```diff
@@ -63,13 +64,21 @@ public StructObjectInspector initialize(ObjectInspector[] args)
     }
     if (!(args[0] instanceof ConstantObjectInspector)) {
       throw new UDFArgumentException(
+          "The first argument to STACK() must be a constant.");
+    }
+    final Object value = ((ConstantObjectInspector) args[0]).getWritableConstantValue();
+    if (value == null) {
+      throw new UDFArgumentException("The first argument of STACK() must not be null.");
+    }
+    if (!(value instanceof IntWritable)) {
```
Comment (SourabhBadhya) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java:73]: Can a long be passed here (LongWritable)?

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java (hunk 4 above)
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java:129]: I added a new statement to `ql/src/test/queries/clientpositive/udtf_stack.q`. The SQL on the current master can't return the correct result.

```
0: jdbc:hive2://hive-hiveserver2:10000/defaul> SELECT STACK(2, 'a', 'b', 'c', 'd', 'e');
...
+-------+-------+-------+
| col0  | col1  | col2  |
+-------+-------+-------+
| a     | b     | c     |
| d     | NULL  | c     |
+-------+-------+-------+
```

This is the correct result.

```
POSTHOOK: query: SELECT STACK(2, 'a', 'b', 'c', 'd', 'e')
POSTHOOK: type: QUERY
POSTHOOK: Input: _dummy_database@_dummy_table
#### A masked pattern was here ####
a	b	c
d	e	NULL
```

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java (hunk 3 above)
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java]: Modified
https://github.com/apache/hive/pull/4938/commits/8b109aa427c65f8443d403f6c3288630a1dfc1ed

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java (hunk 5 above)
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java:72]: Yes, but I believe there is no valid use case for two reasons.
- This UDF can't process N > Integer.MAX_VALUE anyway, as the maximum length of Java's arrays is Integer.MAX_VALUE
- I don't think anyone prefers to write `stack(cast(3 as bigint), a, b, c, ...`

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java (hunk 4 above)
Comment (SourabhBadhya) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java:128]: Nice, glad that a relevant test is added for this bug. Thank you.

**Review by SourabhBadhya — APPROVED** [review]
LGTM, you must probably rebase and retrigger the tests.

**Comment by okumin** [comment]
@SourabhBadhya Thanks for your review! I rebased this branch and confirmed all tests passed 👍

**Comment by okumin** [comment]
Thanks for reviewing and merging this PR 👍
